GH Actions: Nightly builds improvements

marcospereira opened the pull request:
## Changes

This changes some of the existing jobs in a relevant way:

### Before

| Workflow                            | JDK | Scala            | `akka.genjavadoc.enabled` | `akka.test.multi-in-test` | `akka.cluster.assert` |
|:------------------------------------|:----|:-----------------|:--------------------------|:--------------------------|:----------------------|
| `nightly-cluster-metrics-sigar.yml` | 8   | default (`2.13`) | default (`false`)         | default (`true`)          | `on`                  |
| `nightly-jdk8.yml`                  | 8   | default (`2.13`) | default (`false`)         | default (`true`)          | `on`                  |
| `nightly-jdk8-scala213.yml`         | 8   | `2.13`           | default (`false`)         | default (`true`)          | `on`                  |
| `nightly-jdk11.yml`                 | 11  | default (`2.13`) | `true`                    | `false`                   | default (`off`)       |

So, we can see `nightly-jdk8.yml` and `nightly-jdk8-scala213.yml` are actually duplicated (since Scala 2.13 is the default).

### After

We are introducing only one new combination, and removing the duplication of JDK 8/Scala 2.13:

| Job                             | JDK | Scala            | `akka.genjavadoc.enabled`  | `akka.test.multi-in-test` | `akka.cluster.assert` |
|:--------------------------------|:----|:-----------------|:---------------------------|:--------------------------|:----------------------|
| `nightly-cluster-metrics-sigar` | 8   | default (`2.13`) | do not apply*              | default (`true`)          | `on`                  |
| `jdk-nightly-tests`             | 8   | `2.12`           | do not apply*              | `false`                   | `on`                  |
| `jdk-nightly-tests`             | 8   | `2.13`           | do not apply*              | `true`                    | `on`                  |
| `jdk-nightly-tests`             | 11  | `2.12`           | `true`                     | `false`                   | `on`                  |
| `jdk-nightly-tests`             | 11  | `2.13`           | `true`                     | `true`                    | `on`                  |

*`doc` step running only for JDK 11.

The new combination is JDK 11 + Scala 2.12. But there are other changes too:

- JDK 11 combination now runs with `akka.cluster.assert=on`
- JDK 11 + Scala 2.13 now runs with `akka.test.multi-in-test=true`

---

This new setup seems more consistent and easier to maintain in my option. It is also pretty straightforward to change the matrix later and have more handpicked combinations instead of an `n` x `m` matrix with some conditionals (see an example on [Github Action docs](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#using-environment-variables-in-a-matrix)).

patriknw commented on .github/workflows/nightly-builds.yml:
```diff
@@ -0,0 +1,119 @@
+            https://github.com/${{github.repository}}/actions/runs/${{github.run_id}}
+
+  jdk-nightly-tests:
+    name: JDK ${{ matrix.jdkVersion }} / Scala ${{ matrix.scalaVersion }}
```
I guess this is ok for 2x2, but not sure we have to run all combinations when we expand this to let's say 3x3

marcospereira replied:
Yes, I agree. The description is up to date to compare the existing jobs vs. what is proposed by this PR.

patriknw reviewed (commented):
looking good, I guess this is replacing the existing nightly-* workflows and those should be removed?

marcospereira committed:
Some minor build clean ups

- Resolve some warnings due to depracations
- Fix an infinite recursion
- Minor refactoring to use sys.props instead of System.getProperty

marcospereira committed:
Fix GC logging configuration when running with Java 11

Also take the CI environment variable into consideration

marcospereira committed:
Use a single workflow to configure all nightly jobs

patriknw commented on .github/workflows/nightly-builds.yml:
```diff
@@ -0,0 +1,141 @@
+          ## TODO: restore the multi-node tests
+          sbt -jvm-opts .jvmopts-ci \
+            -Dakka.build.scalaVersion=${{ matrix.scalaVersion }} \
+            -Dakka.test.timefactor=10 \
```
We have tried to make tests handle rather random pauses of 2-3 seconds on Jenkins. Default timeout for asserts is 3 seconds. Multiplying that with 10 makes no sense to me. Pauses of 30 seconds from the CI environment is not something we can deal with, and I don't gh actions will have such delays.

I can understand if we want to try with `timefactor=2` but if things are still flaky with that we have other hardening work to do.

marcospereira replied:
> I can understand if we want to try with `timefactor=2` but if things are still flaky with that we have other hardening work to do.

We can go with 2, I agree. I was using the same value we use now for Lagom.

marcospereira replied:
Done in 35dbab7b582d26d4d4cb36675bb1d93186bd5f03.

patriknw reviewed (commented):
looking good, aside from the timefactor

marcospereira reviewed (commented):
Some comments.

marcospereira committed:
Using a timefactor=2

patriknw commented on project/AkkaBuild.scala:
```diff
@@ -191,10 +203,9 @@ object AkkaBuild {
-      else
-        defaults
+      defaults ++ CliOptions.runningOnCi
+        .ifTrue(jvmGCLogOptions(JdkOptions.isJdk11orHigher, JdkOptions.isJdk8))
```
It has been very useful for correlating test failures (timeouts) with GC pauses. Especially now with a new environment we should have it enabled.

patriknw reviewed (approved):
LGTM

patriknw commented:
@marcospereira Created issue https://github.com/akka/akka/issues/30559 for the test failure, but there is also something about the email if you would like to look before merging.

marcospereira committed:
Remove pull_request trigger and add the cron schedule

marcospereira commented:
> but there is also something about the email if you would like to look before merging.

Yeah, that is because the secrets aren't available for pull requests.  This should be good to go now after https://github.com/akka/akka/pull/30551/commits/81afa5bc35a8edd3ca0ee1f5dde3fdae9a362861.

johanandren commented on .github/workflows/nightly-builds.yml:
```diff
@@ -0,0 +1,129 @@
+        # Makes debugging failures easier than looking at the build logs
+        # but we don't need to record test reports for each run.
+        if: ${{ failure() }}
+        uses: marcospereira/action-surefire-report@v1
```
Create an issue to track (so we don't forget)?

marcospereira replied:
Done: https://github.com/akka/akka/issues/30568

johanandren commented on .github/workflows/nightly-builds.yml:
```diff
@@ -0,0 +1,139 @@
+        run: |-
+          sbt -jvm-opts .jvmopts-ci \
+            -Dakka.build.scalaVersion=${{ matrix.scalaVersion }} \
+            publishLocal publishM2
```
This in a matrix - does it mean we could publish a snapshot for Scala 2.13 while 2.12 failed? (Not entirely sure if that is a problem, just something I thought about)

marcospereira replied:
This is publishing local, so we won't have a problem with inconsistent snapshots unless I'm missing something from what you said.

johanandren commented on project/AkkaBuild.scala:
```diff
@@ -21,6 +17,13 @@ import scala.collection.breakOut
+    // CI is the env var defined by Github Actions and Travis:
+    // - https://docs.github.com/en/actions/reference/environment-variables#default-environment-variables
+    // - https://docs.travis-ci.com/user/environment-variables/#default-environment-variables
+    val runningOnCi: CliOption[Boolean] = CliOption("akka.ci-server", sys.env.contains("CI"))
```
I thought it was used in more places but a quick search says only for the GC-logs, so we could probably remove it.

marcospereira replied:
I've declared it because the logic now considers both `akka.ci-server` property and `CI` env var, and I thought that having a `CliOption` makes it slightly clearer. But yeah, I can remove it and check both directly.

johanandren reviewed (approved):
LGTM, with a few questions, good work @marcospereira !

marcospereira commented:
@johanandren got this merged so that we can have these running for nightly builds already. I can follow up on your questions later in future pull requests. :-)
